File: p15.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        if(nums.size() < 3) {
            return {};
        }
        vector<int> last_added = {-1, -1, -1};
        vector<vector<int>> solution;
        sort(nums.begin(), nums.end());
        for(int i = 0; i < nums.size() - 2; ++i) {
            while(i > 0 && i < nums.size() - 3 && (nums[i] == nums[i-1])) {
                ++i;
            }
            if(i > nums.size() - 3) {
                break;
            }
            int sum = -nums[i];
            // two pointers solution for two sum
            int left = i + 1;
            int right = nums.size() - 1;
            while(left < right) {
                if(nums[left] + nums[right] == sum) {
                    if(!(last_added[0] == -sum && last_added[1] == nums[left] && last_added[2] == nums[right])) {
                        solution.push_back({-sum, nums[left], nums[right]});
                        last_added[0] = -sum;
                        last_added[1] = nums[left];
                        last_added[2] = nums[right];
                    }
                }
                if(nums[left] + nums[right] < sum) {
                    ++left;
                } else {
                    --right;
                }
            }
        }
        return solution;
    }
};

File: p15.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        ans = []
        nums.sort()

        for i in range(len(nums)-2):
            if i > 0 and nums[i] == nums[i-1]:
                continue
            left = i+1
            right = len(nums)-1

            while left != right:
                if nums[i] + nums[left] + nums[right] == 0:
                    if len(ans) == 0 or [nums[i], nums[left], nums[right]] != ans[-1]:
                        ans.append([nums[i], nums[left], nums[right]])
                    right -= 1
                elif nums[i] + nums[left] + nums[right] < 0:
                    left += 1
                else:
                    right -= 1

        return ans
